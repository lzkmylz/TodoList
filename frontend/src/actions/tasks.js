import moment from 'moment';
/* 
TASK ACTIONS
Task data structure:
1 task id
2 task start date
3 task expire date
4 task level (0: normal, 1: important, 2: urgent)
5 task title
6 task description
7 isFinish (0 not finish, 1 finish)
*/

// ADD TASK
export const addTask = task => ({
  type: 'ADD_TASK',
  task,
});

export const startAddTask = (taskData = {
  startDate: moment().format(),
  expireDate: moment().format(),
  level: 0,
  title: '',
  description: '',
  isFinish: 0,
}) => {
  return (dispatch, getState) => {
    let headers = {"Content-Type": "application/json"};
    let body = JSON.stringify(taskData);
    return fetch("/api/tasks/", {
      headers,
      method: 'POST',
      body: body,
    }).then(response => {
      response.json().then((responseJson => {
        dispatch(addTask(responseJson));
      }));
    });
  };
};

// REMOVE_TASK
export const removeTask = ({ id }) => ({
  type: 'REMOVE_TASK',
  id,
});

export const startRemoveTask = ({ id }) => {
  return dispatch => {
    let headers = {"Content-Type": "application/json"};

    return fetch(`/api/tasks/${id}/`, {
      headers,
      method: 'DELETE',
    }).then(response => {
        if (response.ok) {
          return dispatch(removeTask({ id }));
        }
      });
  };
};

// EDIT_TASK
export const editTask = (task) => ({
  type: 'EDIT_TASK',
  task,
});

export const startEditTask = (task) => {
  return (dispatch, getState) => {
    let headers = {"Content-Type": "application/json"};
    let data = JSON.stringify({
      startDate: task.startDate,
      expireDate: task.expireDate,
      level: task.level,
      title: task.title,
      description: task.description,
      isFinish: task.isFinish,
    });

    return fetch(`/api/tasks/${task.id}/`, {
      headers,
      method: 'PUT',
      body: data,
    }).then(response => {
      response.json().then((responseJson) => {
        return dispatch(editTask(responseJson));
      });
    });
  };
};

//SET_TASKS
export const setTasks = (tasks) => ({
  type: 'SET_TASKS',
  tasks,
});

export const startSetTasks = () => {
  return (dispatch, getState) => {
    let headers = {"Content-Type": "application/json"};

    return fetch("/api/tasks/", {
      headers,
      method: "GET",
    }).then(response => {
      response.json().then((responseJson) => {
        dispatch(setTasks(responseJson));
      });
    });
  };
};