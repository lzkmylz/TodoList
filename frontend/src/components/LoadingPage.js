import React from 'react'

const LoadingPage = () => (
    <div className="loader">
        <img
          className="loader-image"
          src={process.env.PUBLIC_URL + '/loader.gif'}
          alt="loader-img"
        />
    </div>
);

export default LoadingPage;