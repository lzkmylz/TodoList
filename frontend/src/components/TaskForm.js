import React from 'react';
import moment from 'moment';
import DatePicker from 'antd/lib/date-picker';
import 'antd/lib/date-picker/style/css';
import Form from 'antd/lib/form';
import 'antd/lib/form/style/css';
import Input from 'antd/lib/input';
import 'antd/lib/input/style/css';
import Radio from 'antd/lib/radio';
import 'antd/lib/radio/style/css';
import Button from 'antd/lib/button';
import 'antd/lib/button/style/css';

export class TaskForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      startDate: props.task ? moment(props.task.startDate) : moment(),
      expireDate: props.task ? moment(props.task.expireDate) : null,
      level: props.task ? props.task.level : '0',
      title: props.task ? props.task.title : '',
      description: props.task ? props.task.description : '',
      validDate: true,
    };
  }

  handleSubmit = (e) => {
    e.preventDefault();
    this.props.form.validateFieldsAndScroll((err, values) => {
      if (!err) {
        this.props.onSubmit({
          startDate: moment(values['起始时间']).format(),
          expireDate: moment(values['结束时间']).format(),
          level: values['优先级'],
          title: values['待办事项标题'],
          description: values['待办事项内容'],
          isFinish: "0",
        })
      }
    });
  }

  check = (e, type) => {
    if (type === "checkExpireDateChange") {
      return moment(e).isAfter(moment(this.state.startDate));
    } else if (type === "checkStartDateChange") {
      return moment(this.state.expireDate).isAfter(moment(e));
    }
    return false
  }

  onExpireDateChange = (e) => {
    if (e) {
      this.setState({
        expireDate: e.format(),
        validDate: this.check(e, "checkExpireDateChange")
      });
    } else {
      this.setState({ expireDate: e });
    }
    console.log(this.state);
  }

  onStartDateChange = (e) => {
    if (e) {
      this.setState({
        startDate: e.format(),
        validDate: this.check(e, "checkStartDateChange")
      });
    } else {
      this.setState({ startDate: e });
    }
    console.log(this.state);
  }

  render() {
    const { getFieldDecorator } = this.props.form;
    const FormItem = Form.Item;
    const { TextArea } = Input;
    const formItemLayout = {
      labelCol: {
        xs: { span: 24 },
        sm: { span: 6 },
      },
      wrapperCol: {
        xs: { span: 24 },
        sm: { span: 12 },
      },
    };
    const tailFormItemLayout = {
      wrapperCol: {
        xs: {
          span: 24,
          offset: 0,
        },
        sm: {
          span: 16,
          offset: 6,
        },
      },
    };

    return (
      <Form onSubmit={this.handleSubmit}>
        <FormItem
          {...formItemLayout}
          label="待办事项标题"
        >
          {getFieldDecorator('待办事项标题', {
            initialValue: this.state.title,
            rules: [{
              required: true, message: '必须设定一个标题',
            }],
          })(
            <Input />
          )}
        </FormItem>
        <FormItem
          {...formItemLayout}
          label="待办事项内容"
        >
          {getFieldDecorator('待办事项内容', {
            initialValue: this.state.description,
          })(
            <TextArea />
          )}
        </FormItem>
        <FormItem
          {...formItemLayout}
          label="起始时间"
          validateStatus={this.state.validDate ? "success" : "error"}
          help="起始时间应早于结束时间"
        >
          {getFieldDecorator('起始时间', {
            initialValue: this.state.startDate,
            rules: [{
              required: true, message: '必须选择一个起始时间',
            }],
          })(
            <DatePicker onChange={this.onStartDateChange}/>
          )}
        </FormItem>
        <FormItem
          {...formItemLayout}
          label="结束时间"
          validateStatus={this.state.validDate ? "success" : "error"}
          help="结束时间应晚于起始时间"
        >
          {getFieldDecorator('结束时间', {
            initialValue: this.state.expireDate,
            rules: [{
              required: true, message: '必须选择一个结束时间',
            }],
          })(
            <DatePicker onChange={this.onExpireDateChange}/>
          )}
        </FormItem>
        <FormItem
          {...formItemLayout}
          label="优先级"
        >
          {getFieldDecorator('优先级', {
            initialValue: this.state.level,
            rules: [{
              required: true, message: '必须选择一个优先级',
            }],
          })(
            <Radio.Group buttonStyle="solid">
              <Radio.Button value="0">普通</Radio.Button>
              <Radio.Button value="1">重要</Radio.Button>
              <Radio.Button value="2">紧急</Radio.Button>
            </Radio.Group>
          )}
        </FormItem>
        <FormItem {...tailFormItemLayout}>
          <Button type="primary" htmlType="submit">添加事项</Button>
        </FormItem>
      </Form>
    );
  }
}

export default Form.create()(TaskForm);
