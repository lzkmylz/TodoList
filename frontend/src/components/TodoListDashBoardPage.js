import React from 'react';
import TodoTaskList from './TodoTaskList';
import TodoSummary from './TodoSummary';

const TodoListDashBoardPage = () => (
  <div>
    <TodoSummary />
    <TodoTaskList />
  </div>
);

export default TodoListDashBoardPage;
