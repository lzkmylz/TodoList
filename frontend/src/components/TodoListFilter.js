import React from 'react';
import { connect } from 'react-redux';
import {
  sortByLevel,
  sortByTitletext,
  sortByExpireDate,
  sortByStartDate,
  sortByIsFinish,
} from '../actions/filters';
import Dropdown from 'antd/lib/dropdown';
import 'antd/lib/dropdown/style/css';
import Menu from 'antd/lib/menu';
import 'antd/lib/menu/style/css';
import Button from 'antd/lib/button';
import 'antd/lib/button/style/css';
import Icon from 'antd/lib/icon';
import 'antd/lib/icon/style/css';
import Input from 'antd/lib/input';
import 'antd/lib/input/style/css';
import Switch from 'antd/lib/switch';
import 'antd/lib/switch/style/css';

export class TodoListFilter extends React.Component {

  onTextChange = (e) => {
    this.props.sortByTitletext(e.target.value);
  }

  onChangeSwitch = (e) => {
    this.props.sortByIsFinish(!e);
  }

  onMenuChange = (e) => {
    switch(e.key) {
      case "0":
        this.props.sortByStartDate();
        break;
      case "1":
        this.props.sortByExpireDate();
        break;
      case "2":
        this.props.sortByLevel();
        break;
      default:
        return undefined
    }
  }

  render() {
    const menu = (
      <Menu>
        <Menu.Item key="0" onClick={this.onMenuChange}>
          <span>按创建日期排序</span>
        </Menu.Item>
        <Menu.Item key="1" onClick={this.onMenuChange}>
          <span>按Deadline日期排序</span>
        </Menu.Item>
        <Menu.Item key="2" onClick={this.onMenuChange}>
          <span>按优先级排序</span>
        </Menu.Item>
      </Menu>
    );

    const mapSortBy = {
      level: '按优先级排序',
      expireDate: '按Deadline日期排序',
      startDate: '按创建日期排序',
    }
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <Input 
              size="large"
              placeholder="搜索待办事项"
              type="text"
              value={this.props.filters.titleText} 
              onChange={this.onTextChange}
            />
          </div>
          <div className="input-group__item">
            <Dropdown overlay={menu}>
              <Button size="large">
                {mapSortBy[this.props.filters.sortBy]} <Icon type="down" />
              </Button>
            </Dropdown>
          </div>
          <div className="input-group__item">
            <Switch
              size="default"
              checkedChildren="显示所有待办事项"
              unCheckedChildren="仅显示未完成事项"
              defaultChecked={!this.props.filters.isFinishFilter}
              onChange={this.onChangeSwitch}
            />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
      filters: state.filters
  }
};

const mapDispatchToProps = (dispatch) => ({
  sortByLevel: () => dispatch(sortByLevel()),
  sortByTitletext: (titleText) => dispatch(sortByTitletext(titleText)),
  sortByExpireDate: () => dispatch(sortByExpireDate()),
  sortByStartDate: () => dispatch(sortByStartDate()),
  sortByIsFinish: (isFinishFilter) => dispatch(sortByIsFinish(isFinishFilter))
});

export default connect(mapStateToProps, mapDispatchToProps)(TodoListFilter);
