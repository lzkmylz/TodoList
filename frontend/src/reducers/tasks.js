// TASKS REDUCER
import moment from 'moment';

const taskReducerDefaultState = [{
  id: '1234',
  startDate: moment().subtract(3, 'days'),
  expireDate: moment().add(8, 'days'),
  level: 0,
  title: '迟早要完',
  description: '牙疼',
},{
  id: '2333',
  startDate: moment().subtract(5, 'days'),
  expireDate: moment().add(10, 'days'),
  level: 1,
  title: '毕业',
  description: '新西兰走好',
},{
  id: '666',
  startDate: moment().subtract(10, 'days'),
  expireDate: moment().add(5, 'days'),
  level: 2,
  title: '跑路',
  description: '上了名单了',
}];

const taskReducer = (state = taskReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_TASK':
      return [
        ...state,
        action.task
      ];
    case 'REMOVE_TASK':
      return state.filter(({ id }) => id !== action.id);
    case 'EDIT_TASK':
      return state.map((task) => {
        if (task.id === action.id) {
          return [
            ...task,
            action.updates
          ];
        }
        return true;
      });
    case 'SET_TASKS':
      return action.tasks;
    default:
      return state;
  }
};

export default taskReducer;
