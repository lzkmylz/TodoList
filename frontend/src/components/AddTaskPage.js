import React from 'react';
import { connect } from 'react-redux';
import TaskForm from './TaskForm';
import { startAddTask } from '../actions/tasks';

export class AddTaskPage extends React.Component {

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
              <h1 className="page-header__title">添加待办事项</h1>        
          </div>
        </div>
        <div className="content-container">
          <TaskForm 
          onSubmit={this.onSubmit}
          />
        </div> 
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch) => ({
  startAddTask: (task) => dispatch(startAddTask(task))
});

export default connect(undefined, mapDispatchToProps)(AddTaskPage);
