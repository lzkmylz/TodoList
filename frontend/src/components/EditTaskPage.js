import React from 'react';
import { connect } from 'react-redux';
import TaskForm from './TaskForm';
import { startEditTask, startRemoveTask } from '../actions/tasks';
import Button from 'antd/lib/button';
import 'antd/lib/button/style/css';
import Col from 'antd/lib/col';
import 'antd/lib/col/style/css';

export class EditTaskPage extends React.Component {

  onSubmit = (task) => {
    this.props.startEditTask({
      id: this.props.task.id,
      ...task
    });
    this.props.history.push('/');
  }

  onClick = (e) => {
    this.props.startRemoveTask({ id: this.props.task.id });
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
              <h1 className="page-header__title">编辑待办事项</h1>        
          </div>
        </div>
        <div className="content-container">
          <TaskForm 
            onSubmit={this.onSubmit}
            task={this.props.task}
          />
          <Col xs={{ span:24, offset:0 }} sm={{ span:16, offset:6 }}>
            <Button
              type="primary"
              onClick={this.onClick}
            >
              删除事项
            </Button>
          </Col>  
        </div>
      </div>
    )
  }
}

const mapStateToProps = (state, props) => {
  return {
      task: state.tasks.find((task) => task.id.toString() === props.match.params.id.toString()),
  };
};

const mapDispatchToProps = (dispatch) => ({
  startEditTask: (task) => dispatch(startEditTask(task)),
  startRemoveTask: ({ id }) => dispatch(startRemoveTask({ id })),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditTaskPage);
