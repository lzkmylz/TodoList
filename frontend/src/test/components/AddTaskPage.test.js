import React from 'react';
import { shallow } from 'enzyme';
import { AddTaskPage } from '../../components/AddTaskPage';
import tasks from '../tixtures/tasks';

let startAddTask, history, wrapper;
beforeEach(() => {
  startAddTask = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(<AddTaskPage startAddTask={startAddTask} history={history} />);
})

test('should render AddTaskPage as expect', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle on submit', () => {
  wrapper.find('Form(TaskForm)').prop('onSubmit')(tasks[0]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(startAddTask).toHaveBeenLastCalledWith(tasks[0]);
});
