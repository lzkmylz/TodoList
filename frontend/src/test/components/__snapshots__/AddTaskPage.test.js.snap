// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddTaskPage as expect 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AddTaskPage
    history={
      Object {
        "push": [MockFunction],
      }
    }
    startAddTask={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="page-header"
        >
          <div
            className="content-container"
          >
            <h1
              className="page-header__title"
            >
              添加待办事项
            </h1>
          </div>
        </div>,
        <div
          className="content-container"
        >
          <Form(TaskForm)
            onSubmit={[Function]}
          />
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="content-container"
          >
            <h1
              className="page-header__title"
            >
              添加待办事项
            </h1>
          </div>,
          "className": "page-header",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <h1
              className="page-header__title"
            >
              添加待办事项
            </h1>,
            "className": "content-container",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "添加待办事项",
              "className": "page-header__title",
            },
            "ref": null,
            "rendered": "添加待办事项",
            "type": "h1",
          },
          "type": "div",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Form(TaskForm)
            onSubmit={[Function]}
          />,
          "className": "content-container",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="page-header"
          >
            <div
              className="content-container"
            >
              <h1
                className="page-header__title"
              >
                添加待办事项
              </h1>
            </div>
          </div>,
          <div
            className="content-container"
          >
            <Form(TaskForm)
              onSubmit={[Function]}
            />
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="content-container"
            >
              <h1
                className="page-header__title"
              >
                添加待办事项
              </h1>
            </div>,
            "className": "page-header",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <h1
                className="page-header__title"
              >
                添加待办事项
              </h1>,
              "className": "content-container",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "添加待办事项",
                "className": "page-header__title",
              },
              "ref": null,
              "rendered": "添加待办事项",
              "type": "h1",
            },
            "type": "div",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Form(TaskForm)
              onSubmit={[Function]}
            />,
            "className": "content-container",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "onSubmit": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
